tests/test_checkout.py

```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
+
+    def test_delete_branch(self, written_repo):
+        with pytest.raises(NotImplementedError):
+            # TODO should raise "branch doesn't exist error"
+            written_repo.remove_branch('testbranch')
+        written_repo.create_branch('testbranch')
+        with pytest.raises(NotImplementedError):
+            written_repo.remove_branch('testbranch')
+
+    def test_merge_multiple_checkouts_same_dset(self, written_repo, array5by7):
+        branch1 = written_repo.create_branch('testbranch1')
+        co = written_repo.checkout(write=True, branch_name=branch1)
+        co.datasets['_dset']['1'] = array5by7
+        co.metadata.add('a1', 'b1')
+        co.commit()
+        co.close()
+        branch2 = written_repo.create_branch('testbranch2')
+        co = written_repo.checkout(write=True, branch_name=branch2)
+        co.datasets['_dset']['2'] = array5by7
+        co.metadata.add('a2', 'b2')
+        co.commit()
+        co.close()
+        written_repo.merge('test merge 1', 'master', branch1)
+        written_repo.merge('test merge 2', 'master', branch2)
+        co = written_repo.checkout(branch_name='master')
+        assert len(co.datasets) == 1
+        assert len(co.datasets['_dset']) == 2
+        assert list(co.metadata.keys()) == ['a1', 'a2']
+
+    def test_merge_multiple_checkouts_multiple_dset(self, written_repo, array5by7):
+        branch1 = written_repo.create_branch('testbranch1')
+        co = written_repo.checkout(write=True, branch_name=branch1)
+        co.datasets['_dset']['1'] = array5by7
+        co.commit()
+        co.close()
+        branch2 = written_repo.create_branch('testbranch2')
+        co = written_repo.checkout(write=True, branch_name=branch2)
+        second_dset = co.datasets.init_dataset(name='second_dset', prototype=array5by7)
+        second_dset['1'] = array5by7
+        co.commit()
+        co.close()
+        written_repo.merge('test merge 1', 'master', branch1)
+        written_repo.merge('test merge 2', 'master', branch2)
+        co = written_repo.checkout(branch_name='master')
+        assert len(co.datasets) == 2
+        assert len(co.datasets['_dset']) == 1
+        assert len(co.datasets['second_dset']) == 1
+
+    def test_merge_diverged_conflict(self, written_repo, array5by7):
+        branch1 = written_repo.create_branch('testbranch1')
+        branch2 = written_repo.create_branch('testbranch2')
+        co = written_repo.checkout(write=True, branch_name=branch1)
+        co.datasets['_dset']['1'] = array5by7
+        co.metadata.add('a', 'b')
+        co.commit()
+        co.close()
+        co = written_repo.checkout(write=True, branch_name=branch2)
+        newarray = np.zeros_like(array5by7)
+        co.datasets['_dset']['1'] = newarray
+        co.metadata.add('a', 'c')
+        co.commit()
+        co.close()
+        written_repo.merge('commit message', 'master', branch1)
+        # TODO: This should raise a conflict coz of same key
```

Comment:
I'm not following, can you explain what the bug is a bit better?

Reply:
in both branch1 and branch2, test case has `_dset` that has data on the key `1`. It is essentially a merge conflict while merging but what hangar currently does is overwrite.

Reply:
damn ok. big bug here!

Reply:
created #37